compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -3390,4 +3390,39 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             err.span_label(block.span, "this block is missing a tail expression");
         }
     }
+
+    pub(crate) fn suggest_swapping_lhs_and_rhs(
+        &self,
+        err: &mut Diag<'_>,
+        span: Span,
+        rhs_ty: Ty<'tcx>,
+        lhs_ty: Ty<'tcx>,
+        rhs_expr: &'tcx hir::Expr<'tcx>,
+        lhs_expr: &'tcx hir::Expr<'tcx>,
+        op: hir::BinOp,
+    ) {
+        match op.node {
+            hir::BinOpKind::Eq => {
+                if let Some(partial_eq_def_id) = self.infcx.tcx.lang_items().eq_trait()
+                    && self
+                        .infcx
+                        .type_implements_trait(partial_eq_def_id, [rhs_ty, lhs_ty], self.param_env)
+                        .must_apply_modulo_regions()
+                {
+                    let sm = self.tcx.sess.source_map();
+                    if let Ok(rhs_snippet) = sm.span_to_snippet(rhs_expr.span)
+                        && let Ok(lhs_snippet) = sm.span_to_snippet(lhs_expr.span)
+                    {
+                        err.span_suggestion(
+                                span,
+                                format!("`{rhs_ty}` implements `PartialEq<{lhs_ty}>`; consider swapping the LHS and RHS of the equality"),
+                                format!("{} == {}", rhs_snippet, lhs_snippet),
```

Comment:
before you suggest this, make sure that the entire snippet covered `span` does not
contain any newlines. If there are newlines we shouldn't suggest but just leave the "consider swapping.." as a help.

Add that as a test case

Reply:
I just changed it to this:
```rust
err.note(format!("`{rhs_ty}` implements `PartialEq<{lhs_ty}>`"));
err.multipart_suggestion(
    "consider swapping the equality",
    vec![
        (lhs_expr.span, rhs_snippet),
        (rhs_expr.span, lhs_snippet),
    ],
    Applicability::MaybeIncorrect,
);
```
which seems to do a better job when the entire snippet contains some newlines:
```Shell
error[E0308]: mismatched types
  --> tests/ui/partialeq_suggest_swap.rs:13:5
   |
10 |     4u32
   |     ---- expected because this is `u32`
...
13 |     T(4); //~ mismatched types [E0308]
   |     ^^^^ expected `u32`, found `T`
   |
   = note: `T` implements `PartialEq<u32>`
help: consider swapping the equality
   |
10 ~     T(4)
11 |
12 |     == 
13 ~     4u32; //~ mismatched types [E0308]
   |

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
```
What do you think of this solution?

Reply:
works.